Magic Logins: Forms for requesting and consuming links via "Magic Link" emails

jblz opened the pull request:
# Magic Logins: Forms for requesting and consuming links via "Magic Link" emails

The flow is implemented through 4 routes.

The Email Request Side
----------------

* "Request" page accessible at `/login/send-me-a-link`
  * `POST`s to the `v1.2` API endpoint to request an email w/ a link
<img width="536" alt="screen shot 2017-03-27 at 2 28 54 pm" src="https://cloud.githubusercontent.com/assets/1587282/24371827/171d6a66-12fa-11e7-872e-d6ab5cabf337.png">

* "Check your email" page accessible at `/login/link-was-sent`
  * Displays on successful request for a "Magic Login" email
<img width="536" alt="screen shot 2017-03-27 at 2 36 43 pm" src="https://cloud.githubusercontent.com/assets/1587282/24372135/fa848b5e-12fa-11e7-804e-c75970b8958e.png">


The Clickthrough Handling Side
----------------

* "Clickthrough / Handler" page accessible at `/login/handle-emailed-link`
  * `POST`s to wp-login.php & handles response
<img width="536" alt="screen shot 2017-03-27 at 2 30 02 pm" src="https://cloud.githubusercontent.com/assets/1587282/24372171/1806b35a-12fb-11e7-8ea7-9ad9ea583b6d.png">

* "Invalid / Expired" link page accessible at `/login/link-has-expired`
  * Shows when a user tries to use a link that is malformed or expired
<img width="536" alt="screen shot 2017-03-27 at 2 31 02 pm" src="https://cloud.githubusercontent.com/assets/1587282/24371874/422217d4-12fa-11e7-9a80-6c3ac9e7c533.png">

This not yet enabled in any environment config. Even if you enable the flag, the link does not work without server-side changes.

To test:
* Apply server-side changes (see: D4752-code)
* Run with the feature flag: `ENABLE_FEATURES=magic-login make run`
* Visit http://calypso.localhost:3000/login/send-me-a-link
* Enter an email address for your test account
* Submit
* Check your email
* Clickthrough on the link & follow on-screen instructions

For debugging, paste this in your console:
`localStorage.setItem( 'debug', 'calypso:magic-login' );`

a8ci18n commented on client/auth/emailed-login-link-successfully.jsx:
```diff
@@ -0,0 +1,54 @@
+			<EmptyContent
+				title={ translate( 'Check your Email!' ) }
+				line={ line }
+				action={ translate( 'Back to WordPress.com' ) }
```
Hi! I've found a possible matching string that has already been translated **20** times:
`translate( 'Coming back to WordPress.com?' )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/auth/emailed-login-link-successfully.jsx:
```diff
@@ -0,0 +1,54 @@
+				} )
+				: translate( 'We sent you an email with a magic login link.' ),
+			( <br key="magic-login-line-br" /> ),
+			translate( 'It should arrive within a few minutes. Go click it!', {
```
:information_source: This string is probably long enough that it doesn't need a **context**. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on client/auth/request-login-email-form.jsx:
```diff
@@ -0,0 +1,152 @@
+				</LoggedOutForm>
+				<LoggedOutFormLinks>
+					<LoggedOutFormLinkItem href={ config( 'login_url' ) }>
+						{ translate( 'Enter a password instead' ) }
```
Hi! I've found a possible matching string that has already been translated **35** times:
`translate( 'Enter a Password' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/auth/emailed-login-link-successfully.jsx:
```diff
@@ -0,0 +1,61 @@
+				<EmptyContent
+					title={ translate( 'Check your Email!' ) }
+					line={ line }
+					action={ translate( 'Back to WordPress.com' ) }
```
Hi! I've found a possible matching string that has already been translated **20** times:
`translate( 'Coming back to WordPress.com?' )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/auth/emailed-login-link-expired.jsx:
```diff
@@ -0,0 +1,40 @@
+					illustration={ '/calypso/images/drake/drake-404.svg' }
+					illustrationWidth={ 500 }
+					line={ translate( 'Maybe try resetting your password instead' ) }
+					secondaryAction={ translate( 'Reset my password' ) }
```
Hi! I've found a possible matching string that has already been translated **13** times:
`translate( 'Reset My Password' )` *ES Score: 13*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/try/logged-out-magic-li-req-page)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

jblz replied:
👍

a8ci18n commented on client/auth/emailed-login-link-expired.jsx:
```diff
@@ -0,0 +1,40 @@
+					replaceCurrentLocation={ true }
+				/>
+				<EmptyContent
+					action={ translate( 'Return to WordPress.com' ) }
```
Hi! I've found a possible matching string that has already been translated **16** times:
`translate( 'Return to WordPress Updates page' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

jblz replied:
👎

a8ci18n commented on client/auth/emailed-login-link-successfully.jsx:
```diff
@@ -0,0 +1,61 @@
+					replaceCurrentLocation={ true }
+				/>
+				<EmptyContent
+					action={ translate( 'Back to WordPress.com' ) }
```
Hi! I've found a possible matching string that has already been translated **20** times:
`translate( 'Coming back to WordPress.com?' )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/auth/emailed-login-link-successfully.jsx:
```diff
@@ -0,0 +1,62 @@
+					waitForEmailAddress={ emailAddress }
+				/>
+				<EmptyContent
+					action={ translate( 'Back to WordPress.com' ) }
```
Hi! I've found a possible matching string that has already been translated **20** times:
`translate( 'Coming back to WordPress.com?' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

jblz commented:
@blowery I added you for review here because of our prior discussion about redirecting once logged in.  I wanted to close the loop & let you see how I ended up doing it so you could offer feedback specifically on that portion (ecf0b95)

@scruffian I added you for review because of your ongoing work around login and history w/ this initiative. Any insight you can provide is much appreciated :)

 Thanks in advance to both! :)

scruffian commented on client/auth/emailed-login-link-expired.jsx:
```diff
@@ -0,0 +1,41 @@
+					illustrationWidth={ 500 }
+					line={ translate( 'Maybe try resetting your password instead' ) }
+					secondaryAction={ translate( 'Reset my password' ) }
+					secondaryActionURL={ 'https://wordpress.com/wp-login.php?action=lostpassword' }
```
There is a config value for this (`login_url`).

jblz replied:
Thanks -- will incorporate!

scruffian commented on client/auth/request-login-email-form.jsx:
```diff
@@ -0,0 +1,174 @@
+
+						<LoggedOutFormFooter>
+							<FormButton primary disabled={ this.state.hasSubmitted || ! this.state.isEmailValid }>
+								{ translate( 'Request Email' ) }
```
I know the copy isn't final yet; just mentioning that I find this button text very confusing :)

jblz commented on client/auth/handle-emailed-link-form.jsx:
```diff
@@ -0,0 +1,172 @@
+			)
+		];
+
+		if ( currentUser && currentUser.username ) {
```
If `emailAddress` belongs to `currentUser`, should we just redirect? This would happen if the user clicked the link in the email again after they were already logged in.

jblz commented:
Collected TODOs & concerns in: #12677

mcsf commented on client/components/redirect-when-logged-in/index.jsx:
```diff
@@ -0,0 +1,94 @@
+		debug( this.props.replaceCurrentLocation ? 'replace' : 'assign', this.props.redirectTo );
+		this.props.replaceCurrentLocation
+			? window.location.replace( this.props.redirectTo )
+			: window.location.assign( this.props.redirectTo );
```
Assuming you don't want to use

```js
const methodName = foo ? 'replace' : 'assign';
window.location[ methodName ]( this.props.redirectTo );
```

in order to keep the code easily greppable, this is still a good place for Lodash's `invoke`:

```js
const methodPath = replaceCurrentLocation ? 'location.replace' : 'location.assign';
invoke( window, methodPath, redirectTo );
```

jblz replied:
idk -- I kind of like the style in this PR more than either of those.  Is there a compelling reason to change?

mcsf replied:
Not a compelling one. :) There was syntactic repetition and my brain fixates on that.

mcsf commented on client/components/redirect-when-logged-in/index.jsx:
```diff
@@ -0,0 +1,94 @@
+				return;
+			}
+			this.doTheRedirect();
+			return;
```
Would there be an issue in, instead of immediately redirecting in the same VM frame, deferring it? If not, the whole thing could become flatter:

```js
setTimeout( this.doTheRedirect, delayAtMount || 0 );
return;
```

assuming a previously bound `doTheRedirect`.

mcsf commented on client/auth/controller.js:
```diff
@@ -109,5 +114,49 @@ module.exports = {
+			),
+			document.getElementById( 'primary' ),
+			context.store
+		);
```
Minor: there's a `renderPage` helper which makes this shorter but also untangles the controller from DOM calls.

jblz replied:
#12732 pulls out all the controller / route logic into redux
Thanks for the pointer (I'll try and remember it for next time!)

mcsf commented on client/auth/emailed-login-link-expired.jsx:
```diff
@@ -0,0 +1,47 @@
+	}
+}
+
+// `connect`ing here so `RedirectWhenLoggedIn` can do the same
```
What does this mean? :)

jblz replied:
IIRC, unless this component was `connect`ed, `RedirectWhenLoggedIn` threw an error when trying to do so when mounted as a child.

I'll verify.

mcsf commented on client/auth/handle-emailed-link-form.jsx:
```diff
@@ -0,0 +1,172 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import emailValidator from 'email-validator';
+import request from 'superagent';
```
At some point I'd aim to decouple the request creation (an action) from the view.

jblz replied:
This is taken care of in #12732

mcsf commented on client/auth/request-login-email-form.jsx:
```diff
@@ -0,0 +1,174 @@
+import formState from 'lib/form-state';
+import { getCurrentUser } from 'state/current-user/selectors';
+import { localize } from 'i18n-calypso';
+import wpcom from 'lib/wp';
```
Same as the above file wrt. `superagent`. :)

jblz replied:
Will take care of this in a follow up PR -- #12732 has gotten too big :)

jblz commented on client/components/redirect-when-logged-in/index.jsx:
```diff
@@ -0,0 +1,94 @@
+
+	componentWillUnmount() {
+		debug( 'removing storage event listener' );
+		window.removeEventListener( 'storage', this.storageEventHandler.bind( this ) );
```
Good call -- will fix :)
